### ISSUE TITLE
Rule proposal: disallow implicit casts of value-returning functions to void-returning functions

### ISSUE THREAD

**Issue by inga-lovinde** [typescript-eslint/typescript-eslint#8727]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+new+plugin+rule%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] My proposal specifically checks TypeScript syntax, or it proposes a check that requires type information to be accurate.
- [X] My proposal is not a "formatting rule"; meaning it does not just enforce how code is formatted (whitespace, brace placement, etc).
- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Description

My rule would check that functions returning non-void values cannot be used where functions returning void are expected

### Fail Cases

```TypeScript
[1, 2, 3].forEach(x => x + 1);
```

### Pass Cases

```TypeScript
const results = [1, 2, 3].map(x => x + 1);

[1, 2, 3].forEach(x => console.log(x + 1));

[1, 2, 3].forEach(x => void (x + 1));
```

### Additional Info

For this specific example (and some others) there already is a [`array-callback-return`](https://eslint.org/docs/latest/rules/array-callback-return) rule with `checkForEach` parameter, but it is not type-aware and only covers `.forEach` specifically.

Additionally, typescript-eslint already has a [`no-floating-promises`](https://typescript-eslint.io/rules/no-floating-promises) rule that in many cases prevents using promise-returning functions where void-returning functions are expected.

However, I think that it might be a good idea to disallow any treating of void as a base type, and any implicit conversions of anything to void, unless explicitly specified.

**Comment by bradzacher** [comment]
You didn't mention - what's the motivation here?
What bugs would this rule catch? Can you give specific examples?

**Comment by inga-lovinde** [comment]
One example is that it would catch similar mistakes as `array-callback-return` would, but in a more general and type-aware way.

It's difficult for me to come up with real-world examples, because what I see in a real-world codebase is too complicated to extract just this one thing out of it.

But one oversimplified example would be
```
rows.forEach((row) => {
    if (...something...) {
        return true;
    }

    return false;
});
```
This code is very obviously wrong, it clearly does not do whatever its author intended it to do, yet it is only caught by type-unaware `array-callback-return` rule now because something named `.forEach` is called (and not, say, `lodash.forEachRight`), and because there is an explicit return in the body (which would go away if we extract this function into a separate const):

```
const isValidRow = (row): boolean => { ... };
lodash.forEachRight(rows, isValidRow); // <-- obviously wrong! but there are no linting rules to catch that (even though there are rules to catch `isValidRowAsync` here)
```

Or it could be `emitter.on` instead of `lodash.forEachRight`.
This is a very simplified and artificial example, but I've seen real-world bugs like this in real-world codebases. In most cases, they can be described like this: the author returned a value because they intended it to be used somewhere, because it _should_ have been used somewhere, but they made a mistake and passed that function somewhere that explicitly doesn't care about and disregards return values.
Having a rule that would enforce using the `void` keyword to explicitly communicate that yes, we know that there is a return value and yes, we want to disregard it (or rewriting the code to fix the bug, if there is a bug) would really help.

Also consider this:
```
const f = (a: number): void => a + 1;
```
is a TS type error, but
```
const f: (a: number) => void = a => a+1;
```
is not, even though both expressions are essentially saying the same thing, just in slightly different ways. There is nothing about the second way that would explicitly communicate that we are using it instead of the first one because we don't care about the return value (as e.g. `const f = (a: number): void => void (a + 1);` would communicate). So it would be helpful to have the second form flagged down by linting, just as it is helpful to have the first form flagged down by TS type check; such a rule would help one to catch at least the same kind of bugs as TS compiler catches with the first form. (And when we're doing `.forEach`, or `emitter.on` etc, that's essentially the second form implicitly.)

But yes, I understand that it's not a terribly useful rule.

**Comment by bradzacher** [comment]
I personally see no value in `array-callback-return` in typescript code. It makes sense as a rule in untyped JS because there's nothing to validate return types - I.e. If you write code that uses the result of `forEach` you need a linter to catch that case as you obviously wanted a different iterator.

But in TS attempting to use the result of `forEach` will immediately fail to typecheck because it returns `void`. So no linter is required to alert you of the mistake.

Even still we do have the [`no-confusing-void-expression`](https://typescript-eslint.io/rules/no-confusing-void-expression/) rule which will flag cases of assigning a `void` return somewhere.

In terms of catching the case of an unassigned `forEach` that has an iterator that returns a non-void - I don't think a rule to flag the callback in that case makes much sense. Even if you flag it and the user changes to say `.map` - the code is still broken as `.map` is pure and doesn't mutate the array - so not assigning the value means you've got a useless no-op call.

---

Ultimately I just don't see a whole lot of value in such a lint rule. Being able to assign a non-void-returning function to a location expecting a void returning one is a useful property of TS's void assignment rules.

There is a potential for bugs from this behaviour, sure, but there's also a lot of valid and intentional uses of the behaviour too. A lint rule cannot distinguish between the two and thus would false positive in the latter case. This often leads to worse code as people will try to refactor existing code to return void or assert types in unsafe ways to silence the linter.

There is some niche value to the rule - but I don't think there's broad enough applicability in order for it to be implemented within our general-purpose plugin. As such I'm going to reject this PR. Thanks for filing!
